### DISCUSSION TITLE
Error: Cannot find module 'C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js'

### DISCUSSION THREAD

**Discussion by ghost** [coreybutler/nvm-windows#572] [Support] [answered]
It's uncertain why I am getting this error after running `npm` cli command. Is it because nvm is pointing to that same path?

```bash
internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module 'C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:747:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module 'C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:747:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Comment by ammarsecurity** [comment]
run this in the command line.

SET PATH=C:\Program Files\Nodejs;%PATH%

**Reply by nolanresu** [reply]
if PATH was never touched.

**Reply by Ayman-moh6** [reply]
worked thx

**Reply by Bikramghimire** [reply]
first run command as above it work
but after closing cmd and again same problem

**Reply by arshadmuhd** [reply]
whole day i tried so many solutions nothing worked. after when i commanded this its worked....thanks yaar

**Reply by Prince-thomas03** [reply]
thnks it worked for me

**Reply by msgarga** [reply]
thanks it worked

**Reply by streamlifehd** [reply]
help Pls
Error: Cannot find module '/srv/Python Bot/M-Host'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
    at Function.Module._load (node:internal/modules/cjs/loader:833:27)
    at Object.<anonymous> (/usr/lib/node_modules/pm2/lib/ProcessContainerFork.js:33:23)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

**Reply by coreybutler** [reply]
@streamlifehd - this looks more like a specific module wasn't installed (i.e. was `npm install` executed?).

**Reply by hilfa007** [reply]
Thank you so much..It worked for me

**Reply by myabdi4** [reply]
What the ... how did that work?

**Reply by beingsimple89** [reply]
> run this in the command line.
> 
> SET PATH=C:\Program Files\Nodejs;%PATH%

Thanks this helped me

**Reply by coreybutler** [reply]
Related detail: [PATH order](https://github.com/coreybutler/nvm-windows/wiki/Common-Issues#why-do-i-need-to-uninstall-nodejs-before-installing-nvm-for-windows)

**Reply by luisaraujoc** [reply]
thnks

**Comment by coreybutler** [comment]
For those arriving via Google, search the issues for this problem and review the wiki for answers. There are many reasons why this can happen. The issues have unique responses for different circumstances.

Also see: [Common Issues](https://github.com/coreybutler/nvm-windows/wiki/Common-Issues)

**Comment by vipul-nema** [comment]
![image](https://user-images.githubusercontent.com/24974873/111972427-77812280-8b23-11eb-89b0-1ed28aedfcb5.png)
I have tried all steps. I checked the nodejs folder in nvm root and C:\Program Files\nodejs. both have empty node_modules folder.
If any one can help . Please do.

**Reply by fasatrix** [reply]
+1 same problem

**Reply by Contrapuntist** [reply]
same issue.. it seem specific to newer versions?  At least, version 12 on up no npm eventhough the installation indicates npm is being download and installed unclear where though. 

The directory is empty for me as well. I installed latest node and 14.6.1 (currrent LTS).

**Reply by wbaldanw** [reply]
+1 same problem :-|

**Reply by mgrayum** [reply]
+1 same problem here too

**Reply by coreybutler** [reply]
First, I recommend not using `C:\Program Files` because it is a protected directory which requires different permissions (elevated admin). Some folks, especially independent developers, won't have this problem. It's more common in corporate environments where user accounts are a little more restricted. I tend to use `C:\nodejs` for my `NVM_SYMLINK`.

It's also important to note that `C:\Program Files\Nodejs` is often used by other Node installers as the default. If you have Node installed before installing NVM4W, it will attempt to migrate the existing version for you, but it will fail if the Node process is in use/locked by another user. This is why most version managers suggest uninstalling prior versions of Node before installing the version manager (it's just safer).

The bottom line is 99% of the time this issue manifests as a result of insufficient permissions and/or conflicts with prior versions. I cannot recreate this in a clean Windows environment.

## Update
Due to the fact so many people install NVM4W without knowing whether another installation exists or not (then wondering why it doesn't work), I've create a simple tool to test: [pathfinder.exe](https://github.com/coreybutler/nvm-windows/tree/master/utilities/pathfinder).

**Reply by Xyz39808** [reply]
I have attempted to install a module (twspace-crawler in my case) and doing so appears to delete my copy of npm-cli.js including everything else residing in the same folder. Going to Settings>Apps&Features>Node.js>Repair will consistently restore those files for me.

**Reply by chandanprogrammer** [reply]
Yes! This suggestion really worked.

**Reply by Tuscan-blue** [reply]
Thanks a lot!

**Comment by bruno8moura** [comment]
+1 same problem.

I found that after "nvm install ..." the npm is downloaded as .zip under %NVM_ROOT%/temp but instead of, after download npm package, to copy the npm package to %NODE_ROOT%/node_modules the file .zip is excluded.

In order to work, we need to copy the downloaded package to node_modules manually, however we need to do it before the downloaded npm package is deleted.

**Reply by coreybutler** [reply]
NVM also extracts the zip file, copying relevant files to the NVM_ROOT. The zip file isn't necessary once this is complete.

The most common problem people experience is selecting a target directory their user account doesn't have permission to write to. In other words, NVM4W can download/extract npm, but doesn't have permission to copy the files to the target directory.

**Reply by diputest** [reply]
@coreybutler, does that happen same for mac devices too? If yes, how to give permission to write?

**Reply by coreybutler** [reply]
@diputest I don't know how nvm-sh does it on mac.

**Comment by Araza1994** [comment]
I am getting same issue. Please any one help me

**Reply by Araza1994** [reply]
Worked for me when uninstall node and delete all folders and files related to node and npm in C:\Users\[USER_NAME}\AppData\Roaming folder (npm, npm-cache and capacitor etc) and reinstall node

**Comment by coreybutler** [comment, marked as answer]
Many of those experiencing this issue likely have a space in your username. Make sure you upgrade to NVM 1.1.8, which resolves the space and special character issues.

See the [Common Issues](https://github.com/coreybutler/nvm-windows/wiki/Common-Issues) for more help.

**Reply by futuredeveloper985** [reply]
sorry but this answer does not resolve this problem I have another issue like this 
I can't install npx create-react-app on my computer. it shows cannot find the module.

npx create-react-app myapp
'-------\AppData\Roaming\npm\' is not recognized as an internal or external command,
operable program or batch file.
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module 'C:\Users\----- \node_modules\create-react-app\index.js'       
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)       
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []

**Reply by coreybutler** [reply]
@futuredeveloper985 have you run `nvm install x.x.x` and `nvm use x.x.x` where `x.x.x` is the version of Node you want to use? The error message indicates npm cannot be found. In most cases, this is because `nvm use` hasn't been executed.

Another possible reason is npx may not exist. npx is not available in older versions of npm.

**Reply by Divyansh-singh08** [reply]
i have the same issue how to solve this type of issue  can you plz guide me

**Reply by streamlifehd** [reply]
Help pls

Please create a tokens.txt file and place your token in it!

Process finished with exit code 1

**Comment by canotf** [comment]
可以删除window的环境变量后重启电脑，并重新建立NODE_HOME

**Comment by ShandanaShahid** [comment]
check your all files in folder, may be extra file named SERVER has been created.. Delete it... I faced the same issue... May be this will help..

**Comment by dev8871** [comment]
I was facing the same issue, and the best way to solve this issue is to delete nodejs and reinstall it. Make sure to change the path from program files to somewhere else.

**Reply by SomeDevWeb** [reply]
Yes, that removed the error for me too. Thanks!

**Reply by JihyoPark-github** [reply]
저도 지우고 다시 설치하고 폴더 확인하니까 folder안에 cli.js 파일이 있더니 실행이 되네요 ㅠ 감사합니당

**Reply by pranav-kallikkal** [reply]
It worked. Thanks!

**Reply by Levi-ap** [reply]
yes it worked for me too, tks

**Comment by paolo-lopez-fpuna** [comment]
I solved this problem by removing the old path from node and npm and adding them again.

**Reply by safil-badarudeen** [reply]
i have been trying to solve this problem for whole day.this helped me to solve this problem .thanks brother

**Comment by yosiasz** [comment]
For Windows machines 

1. uninstall all versions of nodejs from your pc using control panel and then cleanup/delete any version you might have out there manually
2.  uninstall nvm using control panel
3. Delete nvm folder from your Roaming folder
4. Install nvm and for your nvm folder use C:\nvm and for your nodejs folder choose c:\nodejs
5. in c:\nvm folder there is a settings.txt file, it should contain the following
root: C:\nvm
path: C:\nodejs
6. I also had another node manager called volta which I had to uninstall


This worked on my laptop.

**Reply by noirlyrik** [reply]
worked for me, thanks!

**Reply by kgkoutis** [reply]
worked for me also

**Reply by phelps-oc** [reply]
I tried the above uninstalling and reinstalling in C:\nvm and node in C:node. I got the same error with the updated directory (Error: Cannot find module 'C:\mnt\c\node\node_modules\npm\bin\npm-cli.js'). I am using node v14.0.0 for an old project. Also in VS code I am using wsl2 and Ubuntu 22. I switched my terminal to use Git Bash and it installed npm fine them. Weird. Maybe this helps someone with a similar setup.

**Reply by coreybutler** [reply]
Please note that as of v1.1.11, there is now a `nvm debug` command that may help your troubleshoot these kinds of issues.

**Comment by ImranRH321** [comment]
![solve](https://user-images.githubusercontent.com/86631487/210506046-4752477c-7f01-4eb0-97a6-cafd503a7e77.png)

**Comment by zhanguoru** [comment]
i do not know what happened. can not run NMP and dispay:Error: Cannot find module 'C:\TEMP\NodeJS\node_modules\npm\bin\npm-cli.js'
I found ..\npm can not find. i copy NPM folder to ..\node_modules from package.
it is running...

**Comment by dlgombert** [comment]
I don't know if this is helpful to anyone, but I tried all of these steps and other things as well and no resolution.  Then I tried to install the same version that was installed on my WSL instance (19.6.1), and for whatever reason it installed perfectly with no issues and now works properly.  There may be something that I did that fixed the underlying issue that I'm not aware of, or maybe this issue only pops up with certain versions of node, or ...  🤷

**Comment by lx78WyY0J5** [comment]
Multiple years later, this issue is still opend.

I followed many steps to debug this... that installing on somewhere without specials char or space 😭


The only fix for me ... dont trust spaces and install somewhere else this crap

**Reply by coreybutler** [reply]
I'm sorry you feel the software is crap. You're welcome to use something else or pay for custom support.

**Reply by lx78WyY0J5** [reply]
> I'm sorry you feel the software is crap. You're welcome to use something else or pay for custom support.

I just found risible that a bug "solved" in v18+ isn't really, and that the default path by installer is not supported 🤣😂
Multiple years, and the only solution is do not add spaces or special char to the path 👎

And BTW fixed myself my src & root path

**Reply by coreybutler** [reply]
The space issue has been resolved for quite some time. Most people falsely assume path issues are about spaces when it is actually non-standard encoding used by Windows to write files in non-English locales. Seeing as that's the only version of Windows I have, testing fixes isn't exactly straightforward.

If you want to contribute something that would be accepted, a PR that supports reading/writing the settings.txt in ASCII or UTF-8 encoding would be a useful fix, assuming it resolves the special character issue for non-US editions of Windows.

**Comment by Leonardll** [comment]
years later i am having the same issues. It is really a nightmare, uninstalling nvm and reinstall it,  updated path, using different node version still gertting the same error message 
node:internal/modules/cjs/loader:1215
  throw err;
  ^

Error: Cannot find module 'is-ci'
Require stack:
- /root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/update-notifier/index.js
- /root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/lib/cli.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1212:15)
    at Module._load (node:internal/modules/cjs/loader:1043:27)
    at Module.require (node:internal/modules/cjs/loader:1298:19)
    at require (node:internal/modules/helpers:182:18)
    at lazy (/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/import-lazy/index.js:2:51)
    at /root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/import-lazy/index.js:10:11
    at new UpdateNotifier (/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/update-notifier/index.js:46:4)
    at module.exports (/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/update-notifier/index.js:183:25)
    at updateCheck (/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/lib/cli.js:54:20)
    at Object.<anonymous> (/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/lib/cli.js:190:1) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/node_modules/update-notifier/index.js',
    '/root/.nvm/versions/node/v20.19.1/lib/node_modules/yo/lib/cli.js'
  ]
}

it is crazy that there is not straight forward fix for this . Installing  a new dependency/library should not be that complicated. 
Blaming it on the user setting is not fair . Something is obviously wrong with this program.

**Reply by coreybutler** [reply]
Sorry for your frustrations. 

It's critically important to understand what NVM for Windows _actually does_. It handles downloads and sets the symlinks. That's all there is, full stop. When you run `node whatever.js` or `npm install`, or anything installed with `npm install -g`, nvm.exe _isn't part of the process at all_. You're running node or npm _directly_. You can look in your task manager to confirm that `node.exe` is the only thing that runs. You won't find `nvm.exe` there.

People are quick to point a finger at NVM4W because they assume it's like other version managers. Most other version managers use shims. A shim is just a wrapper of the same name modifying the command, i.e. a `node` shim wraps the real node executable. With shims, `node <whatever>` and `npm <whatever>` commands would pass through the shim, so there is a possibility they could impact how it runs or cause errors (or catch errors). That's just not how it works in NVM for Windows. NVM for Windows basically says "I set things up, now it's up to Node".
